x/stake/keeper/delegation.go

```diff
@@ -325,14 +321,10 @@ func (k Keeper) unbond(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValA
 	}
 
 	// remove the coins from the validator
-	pool := k.GetPool(ctx)
-	validator, pool, amount = validator.RemoveDelShares(pool, shares)
-
-	k.SetPool(ctx, pool)
+	validator, amount = k.RemoveValidatorTokensAndShares(ctx, validator, shares)
 
-	// update then remove validator if necessary
-	validator = k.UpdateValidator(ctx, validator)
-	if validator.DelegatorShares.IsZero() {
+	if validator.DelegatorShares.IsZero() && validator.Status != sdk.Bonded {
```

Comment:
I'm confused, why is this the correct logic? Is the validator's self bond included as a delegator share?
Also why remove this here instead of in endblock

Reply:
`RemoveValidator` is called because I think it needs to be removed from the new power store ranking.

Reply:
Yes, the self-bond is a usual bond. Has to be removed here since we won't necessarily iterate over it in endblock (we'll only necessarily iterate over it if the validator was bonded).